lib/grandpa/message.go

```diff
@@ -147,152 +96,85 @@ type NeighbourMessage struct {
-	return nil
-}
-
 // CommitMessage represents a network finalisation message
 type CommitMessage struct {
 	Round      uint64
 	SetID      uint64
-	Vote       *Vote
-	Precommits []*Vote
-	AuthData   []*AuthData
+	Vote       Vote
+	Precommits []Vote
+	AuthData   []AuthData
 }
 
-// Decode SCALE decodes the data into a CommitMessage
-func (f *CommitMessage) Decode(r io.Reader) (err error) {
-	f.Round, err = common.ReadUint64(r)
-	if err != nil {
-		return err
-	}
-
-	f.SetID, err = common.ReadUint64(r)
-	if err != nil {
-		return err
-	}
-
-	f.Vote, err = new(Vote).Decode(r)
-	if err != nil {
-		return err
-	}
-
-	sd := &scale.Decoder{Reader: r}
-	numPrecommits, err := sd.Decode(new(big.Int))
-	if err != nil {
-		return err
-	}
-
-	f.Precommits = make([]*Vote, numPrecommits.(*big.Int).Int64())
-	for i := range f.Precommits {
-		f.Precommits[i], err = new(Vote).Decode(r)
-		if err != nil {
-			return err
-		}
-	}
-
-	numAuthData, err := sd.Decode(new(big.Int))
+func (s *Service) newCommitMessage(header *types.Header, round uint64) (*CommitMessage, error) {
+	pcs, err := s.grandpaState.GetPrecommits(round, s.state.setID)
 	if err != nil {
-		return err
-	}
-
-	if numAuthData.(*big.Int).Cmp(numPrecommits.(*big.Int)) != 0 {
-		return ErrPrecommitSignatureMismatch
-	}
-
-	f.AuthData = make([]*AuthData, numAuthData.(*big.Int).Int64())
-	for i := range f.AuthData {
-		f.AuthData[i] = new(AuthData)
-		err = f.AuthData[i].Decode(r)
-		if err != nil {
-			return err
-		}
+		return nil, err
 	}
 
-	return nil
+	precommits, authData := justificationToCompact(pcs)
+	return &CommitMessage{
+		Round:      round,
+		Vote:       *NewVoteFromHeader(header),
+		Precommits: precommits,
+		AuthData:   authData,
+	}, nil
 }
 
-// Type returns commitType
-func (f *CommitMessage) Type() byte {
-	return commitType
-}
+// Index Returns VDT index
+func (f CommitMessage) Index() uint { return 1 }
 
 // ToConsensusMessage converts the CommitMessage into a network-level consensus message
 func (f *CommitMessage) ToConsensusMessage() (*ConsensusMessage, error) {
-	enc, err := scale.Encode(f)
+	msg := newGrandpaMessage()
+	err := msg.Set(*f)
 	if err != nil {
 		return nil, err
 	}
 
-	return &ConsensusMessage{
-		Data: append([]byte{commitType}, enc...),
-	}, nil
-}
-
-func (s *Service) newCommitMessage(header *types.Header, round uint64) (*CommitMessage, error) {
-	pcs, err := s.grandpaState.GetPrecommits(round, s.state.setID)
+	enc, err := scale.Marshal(msg)
 	if err != nil {
 		return nil, err
 	}
 
-	precommits, authData := justificationToCompact(pcs)
-	return &CommitMessage{
-		Round:      round,
-		Vote:       NewVoteFromHeader(header),
-		Precommits: precommits,
-		AuthData:   authData,
+	return &ConsensusMessage{
+		Data: enc,
 	}, nil
 }
 
-func justificationToCompact(just []*SignedVote) ([]*Vote, []*AuthData) {
-	precommits := make([]*Vote, len(just))
-	authData := make([]*AuthData, len(just))
+func justificationToCompact(just []SignedVote) ([]Vote, []AuthData) {
```

Comment:
this function name does not correspond to the action IMO, I guess `splitJustification` or `fromJustificationToVoteAndAuth`

Reply:
That makes sense to me. @noot do you have thoughts on this change?

Reply:
I think this method should be a receiver function of a new type called `type SignedVotes []SignedVote`.

Reply:
I named it this way to match substrate's naming, so I think it's ok to leave as is but up to your judgement